go/src/lib.rs

```diff
@@ -1,9 +1,166 @@
-/**
+/*
  * Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
  */
-use std::ffi::c_void;
+#![deny(unsafe_op_in_unsafe_fn)]
 
+use glide_core::client::Client as GlideClient;
+use glide_core::connection_request;
+use glide_core::errors;
+use glide_core::errors::RequestErrorType;
+use glide_core::ConnectionRequest;
+use protobuf::Message;
+use std::{
+    ffi::{c_void, CString},
+    os::raw::c_char,
+};
+use tokio::runtime::Builder;
+use tokio::runtime::Runtime;
+
+/// Success callback that is called when a Redis command succeeds.
+///
+/// `index_ptr` is a baton-pass back to the caller language to uniquely identify the promise.
+/// `message` is the value returned by the Redis command.
+// TODO: Change message type when implementing command logic
+// TODO: Consider using a single response callback instead of success and failure callbacks
+pub type SuccessCallback = unsafe extern "C" fn(index_ptr: usize, message: *const c_char) -> ();
+
+/// Failure callback that is called when a Redis command fails.
+///
+/// `index_ptr` is a baton-pass back to the caller language to uniquely identify the promise.
+/// `error_message` is the error message returned by Redis for the failed command. It must be manually freed by the caller after this callback is invoked, otherwise a memory leak will occur.
+/// `error_type` is the type of error returned by glide-core, depending on the `RedisError` returned.
+pub type FailureCallback = unsafe extern "C" fn(
+    index_ptr: usize,
+    error_message: *const c_char,
+    error_type: RequestErrorType,
+) -> ();
+
+/// The connection response.
+///
+/// It contains either a connection or an error. It is represented as a struct instead of a union for ease of use in the wrapper language.
+///
+/// The struct is freed using `free_connection_response` to avoid memory leaks. This is achieved by having the external caller invoke this function.
+#[repr(C)]
+pub struct ConnectionResponse {
+    conn_ptr: *const c_void,
+    connection_error_message: *const c_char,
+}
+
+/// The glide client.
+// TODO: Remove allow(dead_code) once connection logic is implemented
+#[allow(dead_code)]
+pub struct Client {
+    client: GlideClient,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+    runtime: Runtime,
+}
+
+fn create_client_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, String> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| err.to_string())?;
+    // TODO: optimize this (e.g. by pinning each go thread to a rust thread)
+    let runtime = Builder::new_current_thread()
```

Comment:
This shouldn't work properly - this uses the calling thread for the runtime, and since the Rust runtime cannot yield control back to the Go runtime, Rust will co-opt the thread. Either use a multi-thread runtime with 1 worker thread, or call this from a new thread.